src/topics/lda_model.cpp

```diff
@@ -16,30 +17,23 @@ lda_model::lda_model(const learn::dataset& docs, std::size_t num_topics)
     /* nothing */
 }
 
-void lda_model::save_doc_topic_distributions(const std::string& filename) const
+void lda_model::save_doc_topic_distributions(std::ostream& stream) const
 {
-    std::ofstream file{filename};
-    for (const auto& doc : docs_)
+    io::packed::write(stream, docs_.size());
+    io::packed::write(stream, num_topics_);
+
+    for (const auto& d : docs_)
     {
-        file << doc.id << "\t";
-        double sum = 0;
+        auto dist = topic_distrbution(doc_id{d.id});
         for (topic_id j{0}; j < num_topics_; ++j)
         {
-            double prob = compute_doc_topic_probability(doc.id, j);
-            if (prob > 0)
-                file << j << ":" << prob << "\t";
-            sum += prob;
+            io::packed::write(stream, dist.counts(j));
         }
-        if (std::abs(sum - 1) > 1e-6)
-            throw std::runtime_error{"invalid probability distribution"};
-        file << "\n";
     }
 }
 
-void lda_model::save_topic_term_distributions(const std::string& filename) const
+void lda_model::save_topic_term_distributions(std::ostream& stream) const
```

Comment:
My suggestion here is to have `top_k` take a scorer parameter, which defines a

```cpp
double operator()(topic_id k, term_id v) const
```

This lets you either pass in a lambda or a functor to the function to define how you want terms to be scored with respect to a topic. So we could do something like

```cpp
auto best = model.top_k(topic, 10, [&](topic_id k, term_id v) {
    return model.term_probability(k, v);
});
```

to get the top-k using just raw probabilities, and something like

```cpp
topics::default_term_scorer scorer{model};
auto best = model.top_k(topic, 10, scorer);
```

to get things with the tfidf-like scorer. (Having it be a separate class allows us to cache the denominator for each vocab item, since it won't change as a function of k.)

I recommend implementing `default_term_scorer` by using log properties and `fastapprox::fastlog` for the logarithms.

```
score(k,v) = \phi_{k,v} (\log \phi_{k,v} - 1/K \sum_j \phi_{j,v})
```

(If anybody can come up with a better name than `default_term_scorer` I'm all ears...)